File: packages-gen/dart/lang/lib/common-sort-topo.dart
sort_edges_build(nodes, edge) {
  var n_from = edge[0];
  var n_to = edge[1];
  if(!nodes.containsKey(n_from)){
    nodes[n_from] = <dynamic, dynamic>{"id":n_from,"links":<dynamic>[]};
  }
  if(!nodes.containsKey(n_to)){
    nodes[n_to] = <dynamic, dynamic>{"id":n_to,"links":<dynamic>[]};
  }
  var links = nodes[n_from]["links"];
  return links.add(n_to);
}

sort_edges_visit(nodes, visited, sorted, id, ancestors) {
  if(true == visited[id]){
    return;
  }
  var node = nodes[id];
  if(null == node){
    throw "Not available: " + id;
  }
  if(null == ancestors){
    ancestors = <dynamic>[];
  }
  ancestors.add(id);
  visited[id] = true;
  var input = node["links"];
  var arr_39661 = input;
  for(var i39662 = 0; i39662 < arr_39661.length; ++i39662){
    var link = arr_39661[i39662];
    sort_edges_visit(
      nodes,
      visited,
      sorted,
      link,
      xt.lang.common_data.arr_clone(ancestors)
    );
  };
  return sorted.insert(0,id);
}

sort_edges(edges) {
  var nodes = <dynamic, dynamic>{};
  var sorted = <dynamic>[];
  var visited = <dynamic, dynamic>{};
  var arr_39683 = edges;
  for(var i39684 = 0; i39684 < arr_39683.length; ++i39684){
    var e = arr_39683[i39684];
    sort_edges_build(nodes,e);
  };
  for(var id in nodes.keys){
    sort_edges_visit(nodes,visited,sorted,id,null);
  };
  return sorted;
}

sort_topo(input) {
  var edges = <dynamic>[];
  var arr_39705 = input;
  for(var i39706 = 0; i39706 < arr_39705.length; ++i39706){
    var link = arr_39705[i39706];
    var root = link[0];
    var deps = link[1];
    var arr_39727 = deps;
    for(var i39728 = 0; i39728 < arr_39727.length; ++i39728){
      var d = arr_39727[i39728];
      edges.add(<dynamic>[root,d]);
    };
  };
  var sorted = sort_edges(edges);
  var out = <dynamic>[];
  for(var i = sorted.length; i > 0; i = (i + -1)){
    out.add(sorted[i + -1]);
  };
  return out;
}

File: src-lang/xt/lang/common_sort_topo.clj
(ns xt.lang.common-sort-topo
  (:require [hara.lang :as l :refer [defspec.xt]]))

(l/script :xtalk
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.common-data :as xtd]]})


;;
;; SORT TOPO
;;

(defn.xt sort-edges-build
  "builds an edge with links"
  {:added "4.0"}
  [nodes edge]
  (var n-from (xt/x:first edge))
  (var n-to   (xt/x:second edge))
  (if (not (xt/x:has-key? nodes n-from))
    (xt/x:set-key nodes n-from {:id n-from
                                :links []}))
  (if (not (xt/x:has-key? nodes n-to))
    (xt/x:set-key nodes n-to {:id n-to
                              :links []}))
  (var links (. nodes [n-from] ["links"]))
  (xt/x:arr-push links n-to))

(defn.xt sort-edges-visit
  "walks over the list of edges"
  {:added "4.0"}
  [nodes visited sorted id ancestors]
  (if (== true (. visited [id]))
    (return))
  (var node (. nodes [id]))
  (if (xt/x:nil? node)
    (xt/x:err (xt/x:cat "Not available: " id)))
  (do (when (xt/x:nil? ancestors)
        (:= ancestors []))
      (xt/x:arr-push ancestors id)
      (xt/x:set-key visited id true)
      (var input (. node ["links"]))
      (xt/for:array [link input]
        (-/sort-edges-visit nodes visited sorted link (xtd/arr-clone ancestors))))
  (xt/x:arr-push-first sorted id))

(defn.xt sort-edges
  "sort edges given a list"
  {:added "4.0"}
  [edges]
  (var nodes   {})
  (var sorted  [])
  (var visited {})
  (xt/for:array [e edges]
    (-/sort-edges-build nodes e))
  (xt/for:object [[id _] nodes]
    (-/sort-edges-visit nodes visited sorted id nil))
  (return sorted))

(defn.xt sort-topo
  "sorts in topological order"
  {:added "4.0"}
  [input]
  (var edges [])
  (xt/for:array [link input]
    (var root (xt/x:first link))
    (var deps (xt/x:second link))
    (xt/for:array [d deps]
      (xt/x:arr-push edges [root d])))
  (var sorted (-/sort-edges edges))
  (var out [])
  (xt/for:index [i [(xt/x:len sorted)
                    (xt/x:offset)
                    -1]]
    (xt/x:arr-push out (xt/x:get-idx sorted (xt/x:offset-rev i))))
  (return out))
